daemon/api_test.go

```diff
@@ -7522,3 +7540,289 @@ func (s *appSuite) TestErrToResponse(c *check.C) {
+}
+
+func (s *apiSuite) TestListSnapshotsListError(c *check.C) {
+	oldList := snapshotList
+	defer func() {
+		snapshotList = oldList
+	}()
+	snapshotList = func(_ context.Context, setID uint64, _ []string) ([]client.SnapshotSet, error) {
+		return nil, errors.New("no")
+	}
+
+	c.Check(snapshotCmd.Path, check.Equals, "/v2/snapshots")
+	req, err := http.NewRequest("GET", "/v2/snapshots", nil)
+	c.Assert(err, check.IsNil)
+
+	rsp := listSnapshots(snapshotCmd, req, nil).(*resp)
+	c.Assert(rsp.Type, check.Equals, ResponseTypeError)
+	c.Check(rsp.Status, check.Equals, 500)
+	c.Check(rsp.Result.(*errorResult).Message, check.Equals, "no")
+}
+
+func (s *apiSuite) TestFormatSnapshotAction(c *check.C) {
+	type table struct {
+		action   snapshotAction
+		expected string
+	}
+	tests := []table{
+		{
+			snapshotAction{SetID: 2, Action: "verb"},
+			`Verb of snapshot set #2`,
+		}, {
+			snapshotAction{SetID: 2, Action: "verb", Snaps: []string{"foo"}},
+			`Verb of snapshot set #2 for snaps "foo"`,
+		}, {
+			snapshotAction{SetID: 2, Action: "verb", Snaps: []string{"foo", "bar"}},
+			`Verb of snapshot set #2 for snaps "foo", "bar"`,
+		}, {
+			snapshotAction{SetID: 2, Action: "verb", Users: []string{"meep"}},
+			`Verb of snapshot set #2 for users "meep"`,
+		}, {
+			snapshotAction{SetID: 2, Action: "verb", Users: []string{"meep", "quux"}},
+			`Verb of snapshot set #2 for users "meep", "quux"`,
+		}, {
+			snapshotAction{SetID: 2, Action: "verb", Users: []string{"meep", "quux"}, Snaps: []string{"foo", "bar"}},
+			`Verb of snapshot set #2 for snaps "foo", "bar" for users "meep", "quux"`,
+		},
+	}
+
+	for _, test := range tests {
+		c.Check(test.action.String(), check.Equals, test.expected)
+	}
+}
+
+func (s *apiSuite) TestChangeSnapshots400(c *check.C) {
+	type table struct{ body, error string }
+	tests := []table{
+		{
+			body:  `"woodchucks`,
+			error: "cannot decode request body into snapshot operation:.*",
+		}, {
+			body:  `{}"woodchucks`,
+			error: "extra content found after snapshot operation",
+		}, {
+			body:  `{}`,
+			error: "snapshot operation requires snapshot set ID",
+		}, {
+			body:  `{"set": 42}`,
+			error: "snapshot operation requires action",
+		}, {
+			body:  `{"set": 42, "action": "carrots"}`,
+			error: `unknown snapshot operation "carrots"`,
+		}, {
+			body:  `{"set": 42, "action": "forget", "users": ["foo"]}`,
+			error: `snapshot "forget" operation cannot specify users`,
+		},
+	}
+
+	for i, test := range tests {
+		comm := check.Commentf("%d:%q", i, test.body)
+		req, err := http.NewRequest("POST", "/v2/snapshots", strings.NewReader(test.body))
+		c.Assert(err, check.IsNil, comm)
+
+		rsp := changeSnapshots(snapshotCmd, req, nil).(*resp)
+		c.Check(rsp.Type, check.Equals, ResponseTypeError, comm)
+		c.Check(rsp.Status, check.Equals, 400, comm)
+		c.Check(rsp.Result.(*errorResult).Message, check.Matches, test.error, comm)
+	}
+}
+
+func (s *apiSuite) TestChangeSnapshots404(c *check.C) {
```

Comment:
If I run this test in isolation it crashes:
```
go test -check.f TestChangeSnapshots404
AppArmor status: apparmor is enabled and all features are available

----------------------------------------------------------------------
PANIC: api_test.go:7696: apiSuite.TestChangeSnapshots404

... Panic: runtime error: invalid memory address or nil pointer dereference (PC=0x459E2A)

/usr/lib/go-1.10/src/runtime/panic.go:502
  in gopanic
/usr/lib/go-1.10/src/runtime/panic.go:63
  in panicmem
/usr/lib/go-1.10/src/runtime/signal_unix.go:388
  in sigpanic
api.go:3027
  in changeSnapshots
api_test.go:7726
  in apiSuite.TestChangeSnapshots404
/usr/lib/go-1.10/src/reflect/value.go:308
  in Value.Call
/usr/lib/go-1.10/src/runtime/asm_amd64.s:2361
  in goexit
OOPS: 0 passed, 1 PANICKED
--- FAIL: Test (0.01s)
FAIL
exit status 1
```

Reply:
WAT

Reply:
figured it out 🤦